src/scheduler.rs

```diff
@@ -96,13 +96,21 @@ impl SchedulerState {
     /// resume will do the right thing and restore the scheduler to the state requested by
     /// the last call.
     pub(crate) async fn pause<'a>(&'_ self, context: Context) -> IoPausedGuard {
-        let mut inner = self.inner.write().await;
-        inner.paused = true;
-        Self::do_stop(inner, &context).await;
-        IoPausedGuard {
-            context,
-            done: false,
+        {
+            let mut inner = self.inner.write().await;
+            inner.paused = true;
+            Self::do_stop(inner, &context).await;
         }
+        let (tx, rx) = oneshot::channel();
+        tokio::spawn(async move {
```

Comment:
Btw, why can't it be spawned in `drop()`?

Reply:
https://docs.rs/tokio/latest/tokio/fn.spawn.html#panics

If you happen to drop this outside of the runtime, which you have effectively no control over, things start going wrong.  This could potentially happen at shutdown time.  By spawning here you have guaranteed correct semantics even when things are being torn down.

If you want to dive deep into this there's a super long essay exploring async drop that I still haven't read myself: https://sabrinajewson.org/blog/async-drop

Reply:
Btw, another approach is to make a generic async function that:
- Pauses IO
- Awaits a future passed to it
- Resumes IO
- Returns the result of the awaited future
This way no async drop is needed.

Reply:
true, callback based APIs are often more awkward to use though